tests/benchq/compilation/test_rbs_with_pauli_tracking.py

```diff
@@ -0,0 +1,574 @@
+            raise Exception(f"other sqs: {sqs}")
+
+    for node, pauli in enumerate(asg["sqp"]):
+        if pauli == jl.I_code:
+            pass
+        elif pauli == jl.X_code_internal:
+            c.x(reg[node])
+        elif pauli == jl.Y_code_internal:
+            c.y(reg[node])
+        elif pauli == jl.Z_code_internal:
+            c.z(reg[node])
+        else:
+            raise Exception(f"other pauli: {pauli}")
+
+
+def measure_node(node, pauli_tracker, c, creg):
+    if pauli_tracker["measurements"][node][0] == jl.I_code:
+        pass
+    elif pauli_tracker["measurements"][node][0] == jl.H_code:
+        c.h(node)
+    elif pauli_tracker["measurements"][node][0] == jl.T_code:
+        c.t(node)
+    elif pauli_tracker["measurements"][node][0] == jl.T_Dagger_code:
+        c.tdg(node)
+    elif pauli_tracker["measurements"][node][0] == jl.RZ_code:
+        c.rz(pauli_tracker["measurements"][node][1], node)
+    else:
+        measurement = pauli_tracker["measurements"][node][0]
+        raise Exception(f"Unknown measurement type: {measurement}")
+
+    c.h(node)
+    c.measure(node, creg[node])
+
+
+def enact_controlled_paulis(
+    x_controls,
+    z_controls,
+    remaining_dag,
+    control,
+    target,
+    control_values,
+    asg,
+    c,
+    creg,
+    node_measured,
+    target_basis,
+):
+    if control in x_controls:
+        remaining_dag[target][0].remove(control)
+        if target in asg["data_nodes"]:
+            c.x(target).c_if(control, control_values[control])
+        else:
+            if target_basis in jl.non_clifford_gate_codes:
+                c.x(target).c_if(control, control_values[control])
+            if target_basis == jl.H_code:
+                if node_measured[target]:
+                    c.x(target).c_if(control, control_values[control])
+                    c.measure(target, creg[target])
+                else:
+                    c.x(target).c_if(control, control_values[control])
+    if control in z_controls:
+        remaining_dag[target][1].remove(control)
+        if (
+            target_basis == jl.I_code
+            or target_basis in jl.non_clifford_gate_codes
+            or target in asg["data_nodes"]
+        ):
+            if node_measured[target]:
+                c.x(target).c_if(control, control_values[control])
+                c.measure(target, creg[target])
+            else:
+                c.z(target).c_if(control, control_values[control])
+
+
+def get_reversed_qiskit_circuit(circuit, data_qubits):
+    shifted_orquestra_circuit = Circuit([])
+    for op in circuit.inverse()._operations:
+        shifted_qubits = (data_qubits[index] for index in op.qubit_indices)
+        shifted_orquestra_circuit += op.gate(*shifted_qubits)
+
+    return export_circuit(QuantumCircuit, shifted_orquestra_circuit)
+
+
+# transformed the counts from the qiskit simulation into a real pdf; some really ugly
+# bit shifting (and similar) is necessary here
+def counts_to_pdf(bits, counts):
```

Comment:
Not something we need to address in this PR, but I just wanted to note that we could probably simplify this logic in the future. For example by using some of the functionality in orquestra-quantum or by doing checks on the counts dict rather than the pdf.

Reply:
Oh yeah, we definitely should...